source/extensions/filters/http/cache/cache_filter.cc

```diff
@@ -128,6 +147,7 @@ Http::FilterDataStatus CacheFilter::encodeData(Buffer::Instance& data, bool end_
     // TODO(toddmgreer): Wait for the cache if necessary.
     insert_->insertBody(
         data, [](bool) {}, end_stream);
+    insert_status_ = CacheInsertStatus::InsertSucceeded;
```

Comment:
In the headers you only set InsertSucceeded if end_stream was true (which as mentioned is still a bit optimistic), but here you're setting it as soon as the first chunk of the body got queued for insertion, which is a bit beyond optimistic. :)
I see in the comment for InsertSucceeded that it's intended to be optimistic about the *last* chunk of the body, so I'm guessing this is supposed to be only happening here if end_stream is true. (I'd suggest also that a test case should be validating that it's *not* set to InsertSucceeded until that phase.)

Reply:
This code makes an assumption that there aren't trailers, which is not a fair assumption in OSS Envoy. Thanks for catching that!